### ISSUE TITLE
wsl2: using custom kernel on arm64

### ISSUE THREAD

**Issue by zlim** [microsoft/WSL#4802]
* Your Windows build number:  (Type `ver` at a Windows Command Prompt)

Microsoft Windows [Version 10.0.19041.1]

* What you're doing and what's happening:  (Copy&paste the full set of _specific_ command-line steps necessary to reproduce the behavior, and their output. Include screen shots if that helps demonstrate the problem.)

Attempting to use a custom kernel, per https://github.com/MicrosoftDocs/WSL/releases/tag/18947.
The following is added to `.wslconfig`:
```
[wsl2]
kernel=C:\\Users\\z\\wsl2\\Image
```
(In PS, confirmed `C:\\Users\\z\\wsl2\\Image` is a valid path.)

* What's wrong / what should be happening instead:

WSL2 doesn't pick up the custom kernel as configured in `.wslconfig`.

* Other experiments:

1. Confirmed .wslconfig is being picked up. With following configuration:
```
[wsl2]
processors=2
```
confirmed WSL2 only shows two processors.

2. Confirmed custom kernel `Image` is valid image. WSL2 boots with custom kernel when `Image` is manually dropped in as replacement of `C:\Windows\System32\lxss\tools\kernel`.

**Comment by benhillis** [comment]
Are you by chance using an ARM64 device? Could you ensure you run WSL.exe --shutdown after adding the entry?

**Comment by zlim** [comment]
> Are you by chance using an ARM64 device?

Yup, Surface Pro X to be specific.

> Could you ensure you run WSL.exe --shutdown after adding the entry?

Yes, the shutdown command is run on every config update.

**Comment by Biswa96** [comment]
Any Event Logs from Windows or `dmesg` from WSL may help to debug this issue. The instructions can be found [here][1].

[1]: https://github.com/microsoft/WSL/blob/master/CONTRIBUTING.md

**Comment by zlim** [comment]
Looking at the logs (via https://github.com/microsoft/WSL/blob/master/CONTRIBUTING.md#start-collecting-logs), the only ErrorCode is 15003. What does this mean? What should I be looking for?

In this case, I removed `C:\Windows\System32\lxss\tools\kernel` so the only available kernel should be the one configured in `.wslconfig`. WSL2 doesn't startup.

Are there specific events to look for related to the handling of custom kernel?

**Comment by zlim** [comment]
The closest thing I could find in Event Viewer related to `kernel` is under `Microsoft-Windows-Hyper-V-Worker/Admin/Hyper-V-Chipset`.

Both success/fail cases show the exact same "kernel" path in the VmBootSourceResult:
```
<DevicePath>AcpiEx(VMBus,0,0)/VenHw(...)/\kernel</DevicePath>
```
I'm guessing there's another mapping that WSL2 would've done before launching the VM.

**Comment by benhillis** [comment]
@zlim - Unfortunately the custom kernel .wslconfig option is not supported on ARM64 yet. We are looking to add support for this in the future.

**Comment by zlim** [comment]
@benhillis - thanks for looking into adding support.

**Comment by Frankdwu** [comment]
Is there any other way to use the custom kernel on ARM64?

**Comment by Biswa96** [comment]
> Is there any other way to use the custom kernel on ARM64?

1. Shutdown any current WSL instances `wsl.exe --shutdown`.
2. Take ownership `C:\Windows\System32\lxss\tools` folder to current user.
3. Rename `C:\Windows\System32\lxss\tools\kernel` to `kernel.old`.
4. Put the new Linux kernel in that path with `kernel` name.

**Comment by Frankdwu** [comment]
> Is there any other way to use the custom kernel on ARM64?
> 
> 
> Shutdown any current WSL instances wsl.exe --shutdown.
> Take ownership C:\Windows\System32\lxss\tools folder to current user.
> Rename C:\Windows\System32\lxss\tools\kernel to kernel.old.
> Put the new Linux kernel in that path with kernel name.

Thanks for your help.
I have tried this step with my custom kernel, but when I am launching my Ubuntu it failed.
I checked the etl logs, it shows one error but I don't find the exact error message.
Attached the error here.

[lxcore_kernel_service_user.zip](https://github.com/microsoft/WSL/files/4197271/lxcore_kernel_service_user.zip)

**Comment by Biswa96** [comment]
I don't know what happens in ARM64 but I was doing those steps at first of WSL2 release (i.e. when custom kernel was not officially supported).

**Comment by Frankdwu** [comment]
> I don't know what happens in ARM64 but I was doing those steps at first of WSL2 release (i.e. when custom kernel was not officially supported).

Thanks for your info.
I checked the errors it is "The operation timed out because a response was not received from the virtual machine or container."
lxcore_service.etl error is "onecore\vm\wsl\lxss\lxcmdlineshared\hvsocket.cpp" and error code is "0x80370109"
lxcore_user.etl exception is "onecore\vm\wsl\lxss\wsl\main.cpp" and "0x80370109".

Maybe this error is ARM64 specific

**Comment by hmartinez82** [comment]
2021 update. The .wslconfig kernel entry is still not being honored, but I was able to run a custom kernel replacing the kernel file under lxss. 

I didn't need a  newer kernel, I just needed the same kernel, but with extra modules and supported features (from https://github.com/rpasek/usbip-wsl2-instructions) . The trick was to copy the current config and apply my chances on top of it with `make menuconfig` aa described in the steps in the link. The kernel version I used is 5.4.72
